### PR TITLE
FilterOrderBy allows `null` and `undefined`

### PR DESCRIPTION
**Pull request by ersimont**
Please fill in this template.

- [X] Make your PR against the `master` branch.
- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes).
- [ ] Run `tsc` without errors.
- [ ] Run `npm run lint package-name` if a `tslint.json` is present.

Select one of these and delete the others:

If changing an existing definition:
- [X] Provide a URL to documentation or source code which provides context for the suggested changes:
  Please see [my SO question](http://stackoverflow.com/q/41604348/1836506). The problem I am addressing: using `strictNullChecks` the current definition does not allow passing in `null` or `undefined`, but Angular does handle it by returning `null` or `undefined`, respectively. There are a few ways to adjust the definition to allow it, each with a drawback as discussed in the SO question, and the comments on the answer. Which is most appropriate?
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "../tslint.json" }`.

### CODE REVIEW

File: angular/index.d.ts
Comment (mhegazy) [inline angular/index.d.ts]: why not:
```ts
<T>(array: T[] | null | undefined , expression: string|((value: T) => any)|(((value: T) => any)|string)[], reverse?: boolean, comparator?: IFilterOrderByComparatorFunc): T[];
```

**Comment by mhegazy** [comment]
@smhc  @ersimont @thorn0 @dnschnur @thetric @aluanhaddad and @diegovilar can you please review?

File: angular/index.d.ts
Comment (thorn0) [inline angular/index.d.ts]: It returns `null` or `undefined` if it's passed `null` or `undefined`.
To see it, go to [this page](https://docs.angularjs.org/api/), open the DevTools console and run:
```js
angular.injector(['ng']).get('orderByFilter')([1,7,5,3]); // [1,3,5,7]
angular.injector(['ng']).get('orderByFilter')(null); // null
```

File: angular/index.d.ts
Comment (thorn0) [inline angular/index.d.ts]: The second parameter (`expression`) should be optional.

**Review by thorn0 — CHANGES_REQUESTED** [review]

File: angular/index.d.ts
Comment (thorn0) [inline angular/index.d.ts]: However, the proposed approach doesn't work. See [playground](http://www.typescriptlang.org/play/#src=interface%20OrderBy%20%7B%0D%0A%20%20%20%20%3CT%2C%20A%20extends%20T%5B%5D%20%7C%20null%20%7C%20undefined%3E(%0D%0A%20%20%20%20%20%20%20%20array%3A%20A%2C%0D%0A%20%20%20%20%20%20%20%20expression%3F%3A%20string%20%7C%20((value%3A%20T)%20%3D%3E%20any)%20%7C%20(((value%3A%20T)%20%3D%3E%20any)%20%7C%20string)%5B%5D%2C%0D%0A%20%20%20%20%20%20%20%20reverse%3F%3A%20boolean%0D%0A%20%20%20%20)%3A%20A%3B%0D%0A%7D%0D%0A%0D%0Alet%20f%3A%20OrderBy%3B%0D%0A%0D%0Alet%20a%20%3D%20f(null%2C%20'z')%3B%20%2F%2F%20%60a%60%20is%20of%20type%20%60any%60%20not%20%60null%60%0D%0A).

Comment (ersimont) [inline angular/index.d.ts]: Oh interesting.  If you turn on `strictNullChecks` in the playground, it starts correctly determining the types instead of making them `any`.  But with the option off, you are correct, this leads to incorrect typing.  Maybe the only way to deal with it in both situations is to list out all 4 signatures. [playground](http://www.typescriptlang.org/play/#src=interface%20OrderBy%20%7B%0D%0A%20%20%20%20%3CT%3E(array%3A%20T%5B%5D%2C%20a_bunch_more_stuff%3F)%3A%20T%5B%5D%3B%0D%0A%20%20%20%20%3CT%3E(array%3A%20T%5B%5D%7Cnull%2C%20a_bunch_more_stuff%3F)%3A%20T%5B%5D%7Cnull%3B%0D%0A%20%20%20%20%3CT%3E(array%3A%20T%5B%5D%7Cundefined%2C%20a_bunch_more_stuff%3F)%3A%20T%5B%5D%7Cundefined%3B%0D%0A%20%20%20%20%3CT%3E(array%3A%20T%5B%5D%7Cnull%7Cundefined%2C%20a_bunch_more_stuff%3F)%3A%20T%5B%5D%7Cnull%7Cundefined%3B%0D%0A%7D%0D%0A%0D%0Alet%20f%3A%20OrderBy%3B%0D%0Alet%20s%3A%20string%5B%5D%3B%0D%0Alet%20n%3A%20string%5B%5D%20%7C%20null%3B%0D%0Alet%20u%3A%20string%5B%5D%20%7C%20undefined%3B%0D%0Alet%20nu%3A%20string%5B%5D%20%7C%20null%20%7C%20undefined%3B%0D%0A%0D%0Alet%20sr%20%3D%20f(s)%3B%0D%0Alet%20nr%20%3D%20f(n)%3B%0D%0Alet%20ur%20%3D%20f(u)%3B%0D%0Alet%20nur%20%3D%20f(nu)%3B%0D%0A)

Comment (smhc) [inline angular/index.d.ts]: I think the existing code is correct as is. It gives the correct types for people using strict null checks, but 'any' for those who aren't, which seems correct behaviour.

Note you can get the correct null return type when typing the paramater as null, even without strict null checks. e.g
[playground](http://www.typescriptlang.org/play/#src=interface%20OrderBy%20{%20%20%20%20<T%2C%20A%20extends%20T[]%20|%20null%20|%20undefined>(%20%20%20%20%20%20%20%20array%3A%20A%2C%20%20%20%20%20%20%20%20expression%3F%3A%20string%20|%20((value%3A%20T)%20%3D>%20any)%20|%20(((value%3A%20T)%20%3D>%20any)%20|%20string)[]%2C%20%20%20%20%20%20%20%20reverse%3F%3A%20boolean%20%20%20%20)%3A%20A%3B}let%20f%3A%20OrderBy%3Blet%20a%20%3D%20f(null%2C%20'z')%3B%20%2F%2F%20`a`%20is%20of%20type%20`any`%20not%20`null`)

File: angular/index.d.ts
Comment (aluanhaddad) [inline angular/index.d.ts]: Yes, angular will supply a default comparer in the absence of `expression`

File: angular/index.d.ts
Comment (thorn0) [inline angular/index.d.ts]: Found a curious edge case. If you try to pass an array of nulls (`[null, null]`) to this function with `strictNullChecks` turned on, you'll get a compilation error.

Comment (aluanhaddad) [inline angular/index.d.ts]: @thorn0 using your example code, sans `--strictNullChecks`
```ts
angular.injector(['ng']).get<IFilterOrderBy>('orderByFilter')([1, 7, 5, 3] as number[] | null) // number[]

angular.injector(['ng']).get<IFilterOrderBy>('orderByFilter')([1, 7, 5, 3] as number[] | undefined) // number[]
```
which as @smhc noted, is correct.

Comment (aluanhaddad) [inline angular/index.d.ts]: > Found a curious edge case. If you try to pass an array of nulls ([null, null]) to this function with strictNullChecks turned on, you'll get a compilation error.

nice catch, and it definitely works fine at runtime, IFF you pass a string predicate for `expression`, but if `expression` is `x =>x.y`, it throws

Comment (aluanhaddad) [inline angular/index.d.ts]:
```js
angular.injector(['ng']).get('orderByFilter')([null, null], 'prop') // [null, null]
```

```js
angular.injector(['ng']).get('orderByFilter')([null, null], x => x.y)
```
results in
> Uncaught TypeError: Cannot read property 'x' of null

Comment (aluanhaddad) [inline angular/index.d.ts]: I found a new wrinkle. When the second type argument is `A extends T[] | null | undefined`, the callback parameter is typed as `{}`, irrespective of `--strictNullChecks`. The type inference use case, IMO the primary one, would be lost if this is merged as is.

Comment (aluanhaddad) [inline angular/index.d.ts]: I am heavily leaning in favor of the following overloads
```ts
<A extends null | undefined | null[] | undefined[]>(array: A, expression?: string | string[], reverse?: boolean, comparator?: IFilterOrderByComparatorFunc): A;
<T>(array: T[], expression?: string | ((value: T) => any) | (((value: T) => any) | string)[], reverse?: boolean, comparator?: IFilterOrderByComparatorFunc): T[];
```
and the addition of this test
```ts
function testOrderByFilter() {
    const ordered = angular.injector(['ng']).get<angular.IFilterOrderBy>('orderByFilter')([1, 2, 3], x => x); // number[]
    const ordered1 = angular.injector(['ng']).get<angular.IFilterOrderBy>('orderByFilter')([null, null], x => x); // null[]
    const orderedNull = angular.injector(['ng']).get<angular.IFilterOrderBy>('orderByFilter')(null); // null
    const orderedUndefined = angular.injector(['ng']).get<angular.IFilterOrderBy>('orderByFilter')([undefined, undefined], x => x); // undefined[]
}
```

Comment (thorn0) [inline angular/index.d.ts]: It's not that we can't use a function as the expression. We just should check `x` for `null` before evaluation `x.y`.
```js
angular.injector(['ng']).get('orderByFilter')([null, null], x => x ? x.y : 0)
```

Comment (aluanhaddad) [inline angular/index.d.ts]: Unfortunately `x` always has the empty type, `{}`, at least on the nightly build. Change should not inside the elements of the Array are nullable.

Comment (aluanhaddad) [inline angular/index.d.ts]: See also my suggestion on how to fix this below.

Comment (thorn0) [inline angular/index.d.ts]: Looks great.

It doesn't work for `orderByFilter(null, x => x)`, but I can't find what can be done about it.

To make the tests stricter, I would add type annotations:
```js
function testOrderByFilter() {
    const orderByFilter = angular.injector(['ng']).get<angular.IFilterOrderBy>('orderByFilter');
    const ordered: number[] = orderByFilter([1, 2, 3], x => x.toFixed()); // number[]
    const ordered1: null[] = orderByFilter([null, null], x => x); // null[]
    const orderedNull: null = orderByFilter(null); // null
    const orderedUndefined: undefined[] = orderByFilter([undefined, undefined], x => x); // undefined[]
}
```

P.S. trying to find a better solution, I've surprisingly discovered that it's valid TypeScript: `interface Foo { <T>(bar: T | T[]): typeof bar; }`. I mean the `typeof` part. Not sure what the use can be though.

Comment (thorn0) [inline angular/index.d.ts]: What about simply this?
```ts
<T extends null | undefined>(array: T, expression?: string | ((value: any) => any) | (((value: any) => any) | string)[], reverse?: boolean, comparator?: IFilterOrderByComparatorFunc): T;
<T>(array: T[], expression?: string | ((value: T) => any) | (((value: T) => any) | string)[], reverse?: boolean, comparator?: IFilterOrderByComparatorFunc): T[];
```

Or should we use something else in place of `((value: any) => any)`? Probably `never` somehow? So that the compiler could tell that it doesn't make sense to pass a function that will never be called?

Comment (thorn0) [inline angular/index.d.ts]: A big problem with these solutions:
```ts
declare function f(): number[] | null;
const x = orderByFilter(f(), x => x ? x * 2 : 0);
```

`Argument of type 'number[] | null' is not assignable to parameter of type 'number[]'.`

**Comment by aluanhaddad** [comment]
@thorn0 I also tried using the `typeof` syntax for the return position, but it doesn't buy us anything in particular here.

As you say, the difficulty comes down to excluding the function type from the `expression` argument whenever `null` or `undefined` is involved. I like your idea of using `never` perhaps that could be the solution.

**Comment by ersimont** [comment]
I'm sure I've missed things, but here is [a gist](https://gist.github.com/ersimont/fe0dcbcaca750349b0dd55d94cff7c8c) will all the options I can see in these comments, each with all the tests I could think of. Paste that into a [playground](http://www.typescriptlang.org/play/) and toggle `strictNullChecks` on and off to see for yourselves.

My comments saying "wrong" are sometimes because of errors and sometimes because of incorrect return types. Of the "incorrect return types" some of those are judgement calls, such as deciding that `{}[]` is never what we want.

Note that the tests whose variables end in `null` (e.g. `o1null`) are probably not worth considering, because who cares about passing in the literal `null`?

In my opinion, `Option1` is the least wrong. However, it would introduce a breaking change in some cases (not sure how many of the other options would, too).

**Comment by aluanhaddad** [comment]
@ersimont 
> I'm sure I've missed things, but here is a gist will all the options I can see in these comments, each with all the tests I could think of. Paste that into a playground and toggle strictNullChecks on and off to see for yourselves.

Thank you very much, firstly for opening this PR, and secondly for capturing this fairly complex discussion into executable form. I am currently reviewing your gist.

**Comment by aluanhaddad** [comment]
@ersimont 
Examples like line 51
```ts
let o1explicit = option1<string>(a, expr); // wrong with & without strict
```
Are not relevant because they are syntax errors. The reason for this is that they explicitly pass a single type argument type to a function which declares two type arguments. So we can rule all such cases out as erroneous. This simplifies things a bit.

File: angular/index.d.ts
Comment (aluanhaddad) [inline angular/index.d.ts]:
```js
angular.injector(['ng']).get('orderByFilter')('A', x => x.toLowerCase()); // ["a"]
```
Whatever is passed to `orderByFilter`, is coerced wrapped into a singleton array if it is not an array before continuing.

I'm starting to thing the whole premise of the filter is unsound.

Comment (thorn0) [inline angular/index.d.ts]:
```js
angular.injector(['ng']).get('orderByFilter')('asd'); // ["a", "d", "s"]
```
It treats strings as 'array-likes'. If we pass something without the `length` property, there will be an exception.

Comment (thorn0) [inline angular/index.d.ts]: Your surprise is caused by the fact that filters weren't designed to be used from JS code. They were created for Angular's extensions of HTML. Their usage looks like this:
```html
    <tr ng-repeat="friend in friends | orderBy:'-age'">
      <td>{{friend.name}}</td>
      <td>{{friend.phone}}</td>
      <td>{{friend.age}}</td>
    </tr>
```
That's why I don't really understand the motivation behind the PRs like this. However, it's not the first one. I suppose people have their reasons why they want these changes.

Comment (aluanhaddad) [inline angular/index.d.ts]: I agree with respect to PR motivation. I think the type of this function doesn't need this level of specificity.

Thank you for explaining the `length` behavior.

That said, even when used from html, treating strings like arrays because they have this property is a bit odd. I've used `ng-repeat` thousands of times and never intentionally passed a string.

Comment (ersimont) [inline angular/index.d.ts]: My motivation was pretty simple and un-nuanced. I am using filter in `.ts` code, and when I changed to start using `strictNullChecks` I got errors. In my case, all of the errors were because I was passing it things that could be `undefined` in addition to an array.

**Comment by ersimont** [comment]
The reason I added things like `line 51` is to show that breaking changes would be introduced to any users who are already explicitly passing the type like that.  It used to take 1 parameter, with option 1 it would take 2.

Re `line 48`: fair enough. I think that's a judgment call whether changing `null` to `any` is correct. I think that is the least important case, because it only comes about when passing in the literal `null`, which nobody should have a reason to do anyway.

**Comment by bowdenk7** [comment]
@aluanhaddad @thorn0 do we think this is good to merge?

**Comment by thorn0** [comment]
This API seems to be too tricky for TS.

Currently, the main problem with the proposed solution is broken type inference:

```ts
declare let arr: string[];
let o1s = orderBy(arr, s => s ? s.toUpperCase() : '');
// Property 'toUpperCase' does not exist on type '{}'.
```

`s` is inferred to be `{}` here. That is type arguments are inferred like this: `orderBy<{}, string[]>`, not `orderBy<string, string[]>` as we'd like them to be.

I tried to use https://github.com/Microsoft/TypeScript/pull/13487 to see whether that feature can help here if I replace `A extends` with `A =`. Doesn't help.

**Comment by ersimont** [comment]
Ah that's another nice test case. I added it to [the gist](https://gist.github.com/ersimont/fe0dcbcaca750349b0dd55d94cff7c8c). It is only a problem in the solution this PR presents. :(

**Comment by thorn0** [comment]
In TS 2.2 RC, the following strange solution works:
```ts
interface OrderBy {
  <T extends Array<any> | null | undefined>(
    array: T,
    expression?: string | ((value: (T & never[])[0]) => any) | 
           Array<string | ((value: (T & never[])[0]) => any)>,
    reverse?: boolean,
    comparator?: IFilterOrderByComparatorFunc
  ): T;
}

declare const orderBy: OrderBy;

// ok1 is string[], x is string
let ok1 = orderBy(['1', '2'], x => x.toUpperCase());
// ok2 is number[], x is number
let ok2 = orderBy([1, 2, 3], [x => x.toFixed()]);
// ok3 in null
let ok3 = orderBy(null);
// ok2 is number[] | null, x is number
let ok4 = orderBy([1, 2] as (number[] | null), x => x.toFixed());
// ok5 is null[], x is null
let ok5 = orderBy([null, null], x => String(x));

// Property 'toUpperCase' does not exist on type 'never'
// let err1 = orderBy(null, s => s.toUpperCase());
// let err2 = orderBy([null], s => s.toUpperCase());
// let err3 = orderBy(undefined, s => s.toFixed()); // Property 'toFixed' does not exist on type 'never'

// let alas1 = orderBy('aasd', s => s.toUpperCase());
```

**Comment by aluanhaddad** [comment]
@thorn0 Awesome discovery! That is a very clever use of `never`. It is also the best solution proposed by far. I say we go with it.

**Comment by thorn0** [comment]
It doesn't work with any other version of TS. Only with 2.2 RC.

**Comment by aluanhaddad** [comment]
@thorn0 well it continues to work through 2.2 RC, now into 2.1 and on through 2.3.0-dev.2017030.
I do wonder _why_ it now works though.

**Comment by minestarks** [comment]
@ersimont I see the motivation. I don't think complicated type arguments are necessary.
This is a good use case for overloads:

```ts

<T>(array: undefined, expression: string|((value: T) => any)|Array<((value: T) => any)|string>, reverse?: boolean, comparator?: IFilterOrderByComparatorFunc): undefined;
<T>(array: null, expression: string|((value: T) => any)|Array<((value: T) => any)|string>, reverse?: boolean, comparator?: IFilterOrderByComparatorFunc): null;
<T>(array: T[], expression: string|((value: T) => any)|Array<((value: T) => any)|string>, reverse?: boolean, comparator?: IFilterOrderByComparatorFunc): T[];
```

Additionally,
-  Is this the only function that has this pattern (of accepting null/undefined and returning the same type)? If similar functions could benefit from the same approach could you update them too?
- Please add a test case that demonstrates your intended usage 
- strictNullChecks should be on in the tsconfig.json in this folder, but it isn't... Could you turn it on so that we catch issues like this one?

**Comment by aluanhaddad** [comment]
@minestarks that overload results in a several incorrect errors including this
```ts
declare const orderBy: angular.IFilterOrderBy;

const orderedNull = orderByFilter(null);
```
which errors even when  `--strictNullChecks` is not enabled.

The complexity of the proposed signatures arose from trying to meet that use case. See the scenario as elaborated in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/14024#discussion_r96529269

> strictNullChecks should be on in the tsconfig.json in this folder, but it isn't... Could you turn it on so that we catch issues like this one?

Definitely this.



> Is this the only function that has this pattern (of accepting null/undefined and returning the same type)? If similar functions could benefit from the same approach could you update them too?

I certainly hope not.

**Comment by thorn0** [comment]
@minestarks see https://github.com/DefinitelyTyped/DefinitelyTyped/pull/14024#discussion_r96734358

**Comment by minestarks** [comment]
@aluanhaddad @thorn0 Hmmm good point. When in doubt, add more overloads?

```ts
    <T>(array: null, expression: string | ((value: T) => any) | (((value: T) => any) | string)[], reverse?: boolean, comparator?: IFilterOrderByComparatorFunc): null;
    <T>(array: undefined, expression: string | ((value: T) => any) | (((value: T) => any) | string)[], reverse?: boolean, comparator?: IFilterOrderByComparatorFunc): undefined;
    <T>(array: T[] | null, expression: string | ((value: T) => any) | (((value: T) => any) | string)[], reverse?: boolean, comparator?: IFilterOrderByComparatorFunc): T[] | null;
    <T>(array: T[] | undefined, expression: string | ((value: T) => any) | (((value: T) => any) | string)[], reverse?: boolean, comparator?: IFilterOrderByComparatorFunc): T[] | undefined;
    <T>(array: T[] | null | undefined, expression: string | ((value: T) => any) | (((value: T) => any) | string)[], reverse?: boolean, comparator?: IFilterOrderByComparatorFunc): T[] | null | undefined;
    <T>(array: T[], expression: string | ((value: T) => any) | (((value: T) => any) | string)[], reverse?: boolean, comparator?: IFilterOrderByComparatorFunc): T[];
```

I any case, I'm not so sure about the value of the first two, so they could potentially be eliminated. What's a use case where you would call this function with a value whose type is known to be ```null``` or ```undefined``` at compile time?

I understand the problem trying to be solved here is more the ```T[] | null``` scenario.

**Comment by thorn0** [comment]
@minestarks Nice. Somehow this idea of adding more overloads didn't cross my mind. I would say it really solves the issue.

A couple of minor points:

1. The last overload (`array: T[]`) should be first, otherwise:

    ```ts
    const foo = orderBy([1]); // foo is number[] | null
    ```

    BTW, if we exclude the first two overloads, we'll get an interesting effect (not sure if this matters):

    ```ts
    const foo = orderBy(null); // foo is {}[] | null
    ```

2. We can finally use `ArrayLike<T>` instead of `T[]` for the first parameter.

**Comment by minestarks** [comment]
@thorn0 Regarding your first point, is there a legitimate case where you would call this function with a variable whose type is known to be ```null``` at compile time? 

In other words, why call ```result = orderBy(null)``` at all? Simpler to set ```result = null```.

If there's no legitimate use case here, there's probably no need to include an overload just for ```null``` (same goes for ```undefined```). The strange typing produced could actually be nice to have, because it alerts you to the fact that you're calling ```orderBy``` unnecessarily.

**Comment by aluanhaddad** [comment]
@minestarks I personally agree with that, I think the behavior is unexpected. But at the same time, we are trying to describe the API more than be proscriptive.

That said, 
```ts
result = orderBy(null);
```
or
```ts
result = orderBy(undefined);
```
would not pass code review in any respectable shop.

**Comment by minestarks** [comment]
Great then. I'll leave the final call to the people who actually use the library. I'll merge once we get a working update people are happy with.

**Comment by thorn0** [comment]
It doesn't have to be exactly `orderBy(null)`. There might be a situation where in `orderBy(someVar, ...)`, `someVar` is inferred to be `null` as a result of narrowing a union type or probably somehow else. I'm totally okay with not worrying about it, just saying there is such a theoretical possibility.

**Comment by ersimont** [comment]
I have fallen behind on the discussion in this PR. If anyone feels like taking it over, please feel free.

**Comment by paulvanbrenk** [comment]
Going to close this for now, since there was no activity for the last 2 weeks and there doesn't seem to be a clear consensus.

**Comment by aluanhaddad** [comment]
@paulvanbrenk Thanks, I think we went down the rabbit hole a bit here 😉

**Pull request closed without merging**
